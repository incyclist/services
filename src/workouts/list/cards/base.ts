import { Card   } from "../../../base/cardlist";
import { PromiseObserver } from "../../../base/types/observer";
import { Workout } from "../../base/model/Workout";

export class BaseCard implements Card<Workout> {
  

    protected visible=true;
    protected initialized

    setInitialized(val: boolean) {
        this.initialized = val
    }

    reset() {        
        this.setInitialized(false)
        this.setVisible(false)
    }
    

    delete(): PromiseObserver<boolean> {
        throw new Error("Method not implemented.");
    }

    getId(): string {
        throw new Error("Method not implemented.");
    }
    getData(): Workout {
        throw new Error("Method not implemented.");
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    setData(_data: Workout) {
        throw new Error("Method not implemented.");
    }

    getCardType() {
        throw new Error("Method not implemented.");
    }
    getDisplayProperties() {
        throw new Error("Method not implemented.");
    }

    isVisible(): boolean {
        return this.visible
    }
    setVisible(visible: boolean) {
        this.visible = visible
    }
    equals(card: Card<Workout>): boolean {
        if (!card)
            return false
        try {
            return card.getCardType()===this.getCardType() && card.getId()===this.getId()
        }
        catch(err) {
            return false
        }

    }

    onMounted(){}
    select() {}
    unselect() {}


}


