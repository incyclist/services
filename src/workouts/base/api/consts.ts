const DEFAULT_WORKOUT_API_PROD = 'https://dlws.incyclist.com/api/v1/routes';

export const WORKOUT_API = 'WORKOUT_API'
export const DEFAULT_WORKOUT_API = DEFAULT_WORKOUT_API_PROD;

export const NO_CACHE = {
    headers: {
        'Cache-Control': 'no-cache',
        'Pragma': 'no-cache',
        'Expires': '0',
    },
};
