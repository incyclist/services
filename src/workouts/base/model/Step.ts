import clone from "../../../utils/clone"
import { valid } from "../../../utils/valid"
import { CurrentStep, DataType, Limit, PowerLimit, StepDefinition } from "./types"

export const STEP_TYPE = {
    STEP : 'step',
    SEGMENT: 'segment'
}

export const POWER_TYPE = {
    WATT : 'watt',
    PCT: 'pct of FTP'
}



export default class Step implements StepDefinition {
    public type: DataType
    public start?:number
    public end?:number
    public duration: number
    public power?: PowerLimit
    public cadence?: Limit
    public hrm?: Limit
    public text: string
    public work: boolean
    public steady: boolean
    public cooldown: boolean

    constructor( opts?, ignoreValidate?:boolean) {

        const {type,start,end,duration, power,cadence,hrm, text, work, steady, cooldown} = opts||{}
        const numVal = (s) => typeof(s)==='string' ? Number(s) : s

        this.type = type ||'step';
        this.start = (valid(start)) ? numVal(start) : undefined;
        this.end = (valid(end)) ? numVal(end) : undefined;
        this.duration = (valid(duration)) ? numVal(duration) :0;
        this.cadence = valid(cadence) ? clone(cadence) : undefined
        this.power = valid(power) ? clone(power) : undefined
        this.hrm = valid(hrm) ? clone(hrm) : undefined
        this.text = text || '';
        this.work = valid(work) ? work : false;
        this.steady = valid(steady) ? steady : true;
        this.cooldown = valid(cooldown) ? cooldown: false;

        if ( ignoreValidate===undefined || ignoreValidate===false)
            this.validate();
    }

    validate() {
        this.validateTiming();
        this.validatePower();
        this.validateCadence();
        this.validateHrm();
    }

    getDuration():number { return this.duration}
    getStart():number { return this.start}
    getEnd():number { return this.end}

    validateTiming():boolean {
        if ( this.start===undefined && this.end===undefined ) {
            throw new Error(`Invalid Step description, start or end needs to be provided `)
        }
        else if ( this.start!==undefined && this.end===undefined) {
            this.end = this.start+ this.duration
            return true;
        }
        else if ( this.start===undefined && this.end!==undefined) {
            this.start = this.end - this.duration
            return true;
        }

        else if ( this.start!==undefined && this.end!==undefined && this.duration===0) {
            this.duration = this.end - this.start
            if ( this.duration<0) throw new Error(`Invalid Step description, duration:${this.duration} `)
            return true;
        }
        else {
            const delta  = Math.abs(this.duration - ( this.end-this.start))
            if ( delta>0) throw new Error(`Invalid step description, duration does not match start and end `)
            return true;
        }
    }

    validateLimit( p:Limit, name:string):boolean {
        if ( valid(p)) {
            if (p.min===undefined && p.max===undefined)
                throw new Error(`Invalid Step description, ${name}: no values specified`)

            if (p.min!==undefined && p.max!==undefined && p.min>p.max )
                throw new Error(`Invalid Step description, min ${name} > max ${name}`)
            if (p.min!==undefined && p.min<0)
                throw new Error(`Invalid Step description, min ${name} <0`)
            if (p.max!==undefined && p.max<0)
                throw new Error(`Invalid Step description, max ${name} <0`)

        }
        return true;
    }

    validatePower():boolean {
        if ( valid(this.power)) {
            const p = this.power;
            if (p.type===undefined) 
                p.type= 'watt';

            return this.validateLimit(p, 'power')
        }
        return true;
    }

    validateCadence() {
        return this.validateLimit(this.cadence, 'cadence')
    }
    validateHrm() {
        return this.validateLimit(this.hrm, 'hrm')
    }

    getLimits( ts:number,includeStepInfo:boolean=false) {
        
        const rv = (limits) => this.getRemainder(limits, includeStepInfo)

        if  (ts>=this.start && ts<=this.end) {
            const duration = this.duration;
            const remaining = this.end-ts;

            if ( this.steady ) {
                 const { power, cadence, hrm, text, work} = this;
                 return rv ({ power, cadence, hrm, text, work,duration,remaining});
            }
            else {
                const { text, work} = this;
                
                const power = this.calc(ts,this.power,includeStepInfo,true);
                const cadence = this.calc(ts,this.cadence,includeStepInfo);
                const hrm = this.calc(ts,this.hrm,includeStepInfo);
                
                return rv({ power, cadence, hrm, text, work,duration, remaining});
            }

        }
        else {
            return undefined
        }
        

    }

    getRemainder ( limits:CurrentStep, includeStepInfo:boolean=false):CurrentStep  {
        if (limits===undefined) return;
        if ( includeStepInfo)
            limits.step = this;
        return limits;
    }


    calc  (ts,limit,includeStepInfo:boolean, isPower=false)  {
        if (!limit)
            return ({});
        const type = isPower ? limit.type : undefined

        const rv = (limits) => this.getRemainder( isPower ? { ...limits, type} : limits, includeStepInfo)

        const part = (ts-this.start) / this.duration;


        if ( limit!==undefined ) {
            if ( limit.max===undefined && limit.min===undefined) {
                return rv({})
            }

            let min,max;
            if ( this.cooldown) {
                if ( limit.max===undefined && limit.min!==undefined) {
                    return rv({min:limit.min})
                } 
                max = limit.min!==undefined ? limit.max-part*(limit.max-limit.min) : limit.max-part*(limit.max)
                min = isPower ? max : limit.min
            }
            else {
                if ( limit.max===undefined && limit.min!==undefined) {
                    return rv({min:limit.min})
                } 
                if ( limit.min===undefined && limit.max!==undefined) {
                    return rv({max:limit.max})
                } 
                max = min = limit.max!==undefined ? part*(limit.max-limit.min)+limit.min : limit.max
                min = isPower ? max : limit.min
            }
            limit = {min,max}
        }
        return rv(limit);
    }

}


