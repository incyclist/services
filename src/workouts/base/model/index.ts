export * from './Workout'
export * from './Segment'
export * from './Step'
export * from './types'