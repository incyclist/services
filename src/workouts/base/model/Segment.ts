import { valid } from '../../../utils/valid';
import { Step } from './Step'
import { CurrentStep, SegmentDefinition, StepDefinition } from './types';

/** @public */
export class Segment extends Step implements SegmentDefinition {

    /** @public the individual steps of this segment */
    steps: Array<Step>
    /** @public number of repetitions */
    repeat: number

    constructor( opts:SegmentDefinition=null, ignoreValidate?:boolean) {
        super(opts,true);

        this.steps = [];
        this.repeat = 1;
        this.type = 'segment';

        if ( valid(opts)) {
            this.init(opts, ignoreValidate);
            if ( ignoreValidate===undefined || ignoreValidate===false)
                this.validate();
        } 
        
    }

    protected init(opts: SegmentDefinition, ignoreValidate: boolean) {
        const start = opts.start || 0;
        this.start = start;
        let stepStart = start;

        if (opts instanceof Step) {
            this.steps = [new Step(opts)];
        }

        else if (opts.steps !== undefined) {
            if (Array.isArray(opts.steps)) {
                opts.steps.forEach(step => {
                    if (step.start === undefined) {
                        step.start = stepStart;
                    }
                    const s = new Step(step);
                    stepStart = s.end;
                    this.steps.push(s);
                });
            }
            else if (ignoreValidate !== true) {
                    throw new Error(`Invalid Segment description, no steps defined`);
            }

        }
        else  if (ignoreValidate === undefined || ignoreValidate === false) {
                throw new Error(`Invalid Segment description, no steps defined`);
        }
        if (opts.repeat !== undefined) this.repeat = opts.repeat;


        this.duration = this.getDuration();
        this.end = this.start + this.duration;
    }

    /* istanbul ignore next */
    validate() {
        this.validateTiming();
    }

    /** @ignore */
    validateTiming() {
        let prev;

        this.steps.forEach( s => {
            s.validateTiming();
            if ( prev!==undefined) {
                const delta1  = Math.abs(prev.end-s.start);
                if (delta1>0) throw new Error(`Invalid Segment description, start & prev end not matching`)
            }
            prev = s;
        });
        return true;
    }

    getDuration() {
        return this.getSingleDuration() * this.repeat;
    }

    getStart() {
        if ( this.start!==undefined)
            return this.start;
        else if ( this.steps.length>0 )
            return this.steps[0].start
        else 
            return undefined;
        
    }

    getEnd() {
        return this.getStart()+this.getDuration();
    }

    getLimits(ts,includeStepInfo=false):CurrentStep {
        const step = this.getStep(ts);
        if ( step===undefined) 
            return undefined;
        
        const segTime = ts-this.getStart();
        const stepDuration = this.getSingleDuration();

        let part = (segTime % stepDuration);
        if (part===stepDuration)
            part=0;

        part +=this.getStart()
        if ( part===this.getStart() && segTime>0 && segTime<stepDuration)
            part+=stepDuration;
        
        
        return step.getLimits(part,includeStepInfo);
    }

    protected prepareNext(json:StepDefinition|SegmentDefinition) {
        /* istanbul ignore next */
        if ( json===undefined)
            return;
        
        const empty = this.steps.length===0;
        if (json.start===undefined && json.end===undefined) {
            json.start = !empty ? this.steps[this.steps.length-1].end : 0;
        }
        if ( empty) {
            this.start = 0;
        }
    }

    // push methos is in principle protected, but in order to support testing it was declared public
    /** @ignore */
    push(s) {
        if ( s===undefined)
            return;

        this.prepareNext(s);
        let step;
        if ( s instanceof Step || s instanceof Segment)
            step = s;
        else 
            step = new Step(s)
        this.steps.push(step);
        this.duration += step.getDuration();
        this.end = this.start + this.duration;

        this.validate();
        
    }


    protected getSingleDuration() {
        if (this.steps.length===0) return 0;
        return this.steps.map(s=>s.duration).reduce( (a,b) => a+b,0 );
    }


    protected getStep(ts:number):Step {
        if  (ts>=this.getStart() && ts<this.getEnd()) {
            const segTime = ts-this.getStart();
            let part = (segTime % this.getSingleDuration());
            if (part===this.getSingleDuration())
                part=0;
            part +=this.getStart()
            if ( part===this.getStart() && segTime>0 && segTime<this.getSingleDuration())
                part+=this.getSingleDuration();
            const found = this.steps.find( s => {
                return (part>=s.start && part<s.end) 
            })
            return found;
        }
        
    }

}

