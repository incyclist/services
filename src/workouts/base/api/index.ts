import {EventLogger} from 'gd-eventlog'
import { AxiosInstance } from "axios";
import { IncyclistRestApiClient, RestApiClient } from '../../../api';
import { useUserSettings } from '../../../settings';
import { Singleton } from '../../../base/types';
import { DEFAULT_WORKOUT_API, WORKOUT_API } from './consts';
import { Plan, Workout } from '../model/Workout';

@Singleton
export class IncyclistWorkoutsApi { 

    protected logger: EventLogger
    protected api: AxiosInstance

    constructor() {
        this.logger = new EventLogger('WorkoutsApi')
    }

    protected logError( err:Error, fn:string, logInfo?) {
        const args = logInfo || {}
        this.logger.logEvent( {message:'Error', error:err.message, fn, ...args})
    }

    protected getApi():AxiosInstance {
        if (!this.api) {
            this.api = IncyclistRestApiClient.getClient()
            return this.api
        }
        return this.api
    }


    protected getBaseUrl():string {
        
        return useUserSettings().get(WORKOUT_API,DEFAULT_WORKOUT_API)
    }


    async getWorkouts():Promise<Array<Workout|Plan>> {
        try {
            const res = await this._get( '/' )              
            return res.data;           
        }
        catch(err) {
            return undefined
        }

    }

    async getWorkout(id:string):Promise<Workout|Plan> {
        try {
            const res = await this._get( `/${id}` )              
            return res.data;           
        }
        catch(err) {
            return undefined
        }
        
    }

    async reload() {
        try {
            this._get( `/reload` )                   
        }
        catch(err) {
            this.logError(err,'reload')
            return undefined
        }

    }

    protected async _get(url:string, ...args) {
        const api = this.getApi()
        const baseUrl = this.getBaseUrl()
        const request = baseUrl+url
        console.log(request)

        return await api.get( request, ...args )       
    }


}




