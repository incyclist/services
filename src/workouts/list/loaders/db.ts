import { JSONObject, JsonRepository } from "../../../api"
import { Singleton } from "../../../base/types"
import {  Observer, PromiseObserver } from "../../../base/types/observer"
import { Loader } from "./loader"
import { Plan, Workout } from "../../base/model/Workout"
import { EventLogger } from "gd-eventlog"
import { waitNextTick } from "../../../utils"

@Singleton
export class WorkoutsDbLoader extends Loader{
   
    protected saveObserver: PromiseObserver<void>
    protected repo: JsonRepository
    protected workouts: Array<Workout|Plan>
    protected tsLastWrite:number
    protected isDirty:boolean
    protected workoutsSaveObserver:{ [index:string]:PromiseObserver<void>} = {}
    protected logger:EventLogger
    

    constructor () {
        super()
        this.workouts = []     
        this.isDirty = true;
        this.logger = new EventLogger('WorkoutsDB')        
    }

    load(): Observer {       
        if (this.loadObserver)
            return this.loadObserver;

        this.loadObserver = new Observer();
        this._load();
        return this.loadObserver;

    }

    stopLoad() {
        delete this.loadObserver;
    }



    protected getRepo() {
        if (!this.repo)
            this.repo  =JsonRepository.create('workouts')
        return this.repo
    }

    protected async _load() {

        const workouts = await this.getRepo().read('db') 
        if (workouts) {

            const items = workouts as unknown as Array<Workout|Plan>
            items.forEach( item=> {
                const workout = new Workout(item)
                this.workouts.push(workout )
                this.emitAdded(workout)
            })
            
        }

        this.emitDone()
      
    }


    async save(workout:Workout|Plan):Promise<void> {
        const stringify = (json) => { try {JSON.stringify(json)} catch {/* */}}

        let prev
        const idx = this.workouts.findIndex( d=> d.id===workout.id)

        if (idx===-1)
            this.workouts.push( workout)
        else { 
            this.workouts[idx] = workout
            
        }

        const changed = !prev || stringify(this.workouts)!==prev
    
        if (changed) {
            this.isDirty = true;
            this.write()
        }
    }

    async delete(workout:Workout|Plan):Promise<void> {
        const id = workout.id
        const idx = this.workouts.findIndex( d=> d.id===id)
        if (idx==-1)
            throw new Error('workout not found')

        this.workouts.splice(idx,1)

        this.isDirty = true;
        await this.write()
    }


    protected async writeRepo() {
        // avoid concurrent updates
        if (this.saveObserver)
            await this.saveObserver.wait()

        const save = async ():Promise<void>=> {
            try {
                await this.repo.write('db',this.workouts as unknown as JSONObject)
            }
            catch(err) {
                this.logger.logEvent({message:'could not safe repo',error:err.message })
            }
        }
       
        this.saveObserver = new PromiseObserver( save())
        await this.saveObserver.start()
        process.nextTick( ()=> {delete this.saveObserver})

    }


    protected write() {

        if (this.isDirty && (this.tsLastWrite===undefined || Date.now()-this.tsLastWrite>=1000)) {
            this.isDirty = false;
            this.tsLastWrite = Date.now()            
            this.writeRepo()
        }

        if (this.isDirty && Date.now()-this.tsLastWrite<1000) {
            setTimeout( ()=>{this.write()},  this.tsLastWrite+1000-Date.now())
        }

    }

    protected emitUpdated(workout: Workout|Plan) {
        if (this.loadObserver)
            this.loadObserver.emit('workout.updated',workout)
    }
    protected emitAdded(workout: Workout|Plan) {
        if (this.loadObserver)
            this.loadObserver.emit('workout.added',workout)
    }


    protected emitDone() {
        if (this.loadObserver)
            this.loadObserver.emit('done')
        
        waitNextTick().then(()=>{
            this.loadObserver.reset()
            delete this.loadObserver
        })
    }



}
