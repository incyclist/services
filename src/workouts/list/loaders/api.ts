import { Observer } from "../../../base/types/observer";
import { waitNextTick } from "../../../utils";
import { IncyclistWorkoutsApi} from "../../base/api";
import { PlanDefinition } from "../../base/model";
import { Workout,Plan } from "../../base/model/Workout";
import { WorkoutsDbLoader } from "./db";
import { Loader } from "./loader";


export class WorkoutsApiLoader extends Loader {

    protected api: IncyclistWorkoutsApi

    constructor () {
        super()
        this.api = new IncyclistWorkoutsApi() // singelton
    }


    load():Observer {
        // no concurrent loads
        if (this.loadObserver)
            return this.loadObserver

        this.loadObserver = new Observer()

        this.api.getWorkouts().catch().then (async (res:Array<Workout|Plan>) => {
            this.workouts = res

            res.forEach( record=> {

                let data = record;

                if (record.type==='workout') data = new Workout(record)
                if (record.type==='plan') data = new Plan(record as PlanDefinition)

                const id = data.id
                const fromDB = this.getFromDB(id)
                if (!fromDB || fromDB.hash!==data.hash) {
                    this.loadObserver.emit( fromDB?  'workout.updated' : 'workout.added', data)
                    this.save(record)
                }
            })

            this.loadObserver.emit('done')
            await waitNextTick()
            delete this.loadObserver
           
        });

        return this.loadObserver
        
    }


    protected getFromDB(id:string) {
        const db = new WorkoutsDbLoader()
        return db.get(id)

    }

    async save(record:Workout|Plan):Promise<void> {
        const db = new WorkoutsDbLoader()
        await db.save(record)
    }



 
}