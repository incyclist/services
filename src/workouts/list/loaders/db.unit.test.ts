

//import routesData from '../../../../__tests__/data/db/routes.json';
import repoData from '../../../../__tests__/data/workouts/db.json'


import { JSONObject, JsonRepository } from '../../../api';
import { WorkoutsDbLoader } from './db';
import { Workout } from '../../base/model';

class MockRepository extends JsonRepository {
    static create(repoName:string):JsonRepository {
        return super.create(repoName)
    }
    static reset() {
        JsonRepository._instances = {}
        
    }
}


async function dbTest(loader: WorkoutsDbLoader, workouts: Workout[],id:string) {
    const repo = MockRepository.create('workouts');


    repo['id'] = `${id}`

    repo.read = jest.fn(async (file) => {
        if (file === 'db') {
            const workout = repoData.find(r => r.id === id);
            return [workout] as unknown as JSONObject;
        }
        return null as unknown as JSONObject;
    });

    repo.write = jest.fn();


    
    await new Promise(done => {
        loader.load()
            .on('workout.added', workout => { workouts.push(workout); })
            .on('workout.updated', done)
            .on('done', () => {                 
                setTimeout(done, 100); 
            });           
    });

    console.log('done - observer:',loader['loadObserver'])
    
    repo.read = jest.fn()

}






describe('WorkoutDBLoader',()=>{
    describe('load',()=>{

  
        let loader:WorkoutsDbLoader;
        let loaderObj 
        beforeEach( ()=>{
            loader = loaderObj = new WorkoutsDbLoader()                               
        })

        afterEach( ()=>{
            // cleanup Singletons
            loaderObj.reset()            
            MockRepository.reset()
        })


        test('Workout File from DB',async ()=>{
            const workouts:Array<Workout> = []

            const src = repoData.find( w=>w.id==='a1d94b31399d53b0a26d46baa7114884')

            await dbTest(loader, workouts,'a1d94b31399d53b0a26d46baa7114884');            
            expect(workouts.length).toBe(1)

            const wo = workouts[0]
            expect(wo.name).toBe('4. Threshold Under/Overs')
            expect(wo.steps.map(s=>s.type).join(',')).toEqual(src?.steps.map(s=>s.type).join(',')  )
            expect(wo.duration).toBe(49*60)

            loaderObj = {...loader}
            expect(loaderObj['loadObserver'] ).toBeUndefined()

        })

        test('Steps only',async ()=>{
            const workouts:Array<Workout> = []

            const src = repoData.find( w=>w.id==='b5e66d8664c110f67cf5dea67809af00')
            
            await dbTest(loader, workouts,'b5e66d8664c110f67cf5dea67809af00');            
            expect(workouts.length).toBe(1)

            const wo = workouts[0]
            expect(wo.name).toBe('6. FTP Boost')
            expect(wo.steps.map(s=>s.type).join(',')).toEqual(src?.steps.map(s=>s.type).join(',')  )
            expect(wo.duration).toBe(3570)
            expect(wo.getDuration()).toBe(3570)

            loaderObj = {...loader}
            expect(loaderObj['loadObserver'] ).toBeUndefined()

        })



    })

})
