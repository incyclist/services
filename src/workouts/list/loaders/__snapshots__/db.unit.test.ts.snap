// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkoutDBLoader load Workout File from DB 1`] = `
[
  Step {
    "cadence": undefined,
    "cooldown": false,
    "duration": 600,
    "end": 600,
    "hrm": undefined,
    "power": {
      "max": 80,
      "min": 50,
      "type": "pct of FTP",
    },
    "start": 0,
    "steady": false,
    "text": "",
    "type": "step",
    "work": false,
  },
  Step {
    "cadence": undefined,
    "cooldown": false,
    "duration": 60,
    "end": 660,
    "hrm": undefined,
    "power": {
      "max": 95,
      "min": 80,
      "type": "pct of FTP",
    },
    "start": 600,
    "steady": false,
    "text": "",
    "type": "step",
    "work": true,
  },
  Step {
    "cadence": undefined,
    "cooldown": false,
    "duration": 180,
    "end": 840,
    "hrm": undefined,
    "power": {
      "max": 55.00000000000001,
      "min": 55.00000000000001,
      "type": "pct of FTP",
    },
    "start": 660,
    "steady": true,
    "text": "",
    "type": "step",
    "work": true,
  },
  Step {
    "cadence": undefined,
    "cooldown": false,
    "duration": 720,
    "end": 1560,
    "hrm": undefined,
    "power": undefined,
    "start": 840,
    "steady": true,
    "text": "",
    "type": "step",
    "work": false,
  },
  Step {
    "cadence": undefined,
    "cooldown": false,
    "duration": 360,
    "end": 1920,
    "hrm": undefined,
    "power": {
      "max": 55.00000000000001,
      "min": 55.00000000000001,
      "type": "pct of FTP",
    },
    "start": 1560,
    "steady": true,
    "text": "",
    "type": "step",
    "work": true,
  },
  Step {
    "cadence": undefined,
    "cooldown": false,
    "duration": 720,
    "end": 2640,
    "hrm": undefined,
    "power": undefined,
    "start": 1920,
    "steady": true,
    "text": "",
    "type": "step",
    "work": false,
  },
  Step {
    "cadence": undefined,
    "cooldown": true,
    "duration": 300,
    "end": 2940,
    "hrm": undefined,
    "power": {
      "max": 65,
      "min": 45,
      "type": "pct of FTP",
    },
    "start": 2640,
    "steady": false,
    "text": "",
    "type": "step",
    "work": false,
  },
]
`;
